app/lib/metrics_service.rb

```diff
@@ -0,0 +1,108 @@
+# Service to record events to Mixpanel via Bard
+#
+# Mixpanel is a business analytics service for tracking user interactions.
+# Bard is a DSP service that mediates writes to Mixpanel.
+#
+# Most logging to Mixpanel is done in JavaScript; see metrics-api.js and
+# scp-api-metrics.js in /app/javascript/lib/.  But some logging is done
+# in Rails to make it easy to analyze metrics around long-running events
+# triggered by user interactions -- like upload and ingest -- or special
+# identity resolution.
+
+class MetricsService
+
+  BARD_ROOT = Rails.application.config.bard_host_url
+
+  def self.post_to_bard(params, user)
+    params.merge!({
+      method: 'POST'
+    })
+    begin
+      Rails.logger.info "#{Time.zone.now}: Posting to Mixpanel.  Params: #{params}"
+      # Uncomment line below to get known-good test data
+      # puts "Posting to Mixpanel.  Params: #{params}"
+      RestClient::Request.execute(params)
+    rescue RestClient::ExceptionWithResponse => e
+      Rails.logger.error "#{Time.zone.now}: Bard error in call to #{params[:url]}: #{e.message}"
+      # Rails.logger.error e.to_yaml
+      ErrorTracker.report_exception(e, user, params)
+    end
+  end
+
+  def self.get_default_headers(user)
```

Comment:
I would rewrite this as:

```
{
  'Authorization': "Bearer #{user.valid_access_token['access_token']}",
  'Content-Type': 'application/json'
}
```

We don't need to call `return` as the default Ruby behavior is to return the last object in memory.  Also, calling `valid_access_token` will ensure that we have a good OAuth token, in case it has expired (these tokens are not renewed automatically with a user's continued session, so we have to manually refresh them).

Reply:
Good points, done in 6265f13.